### PR TITLE
[Keymap] Add DZ60, Levinson keymaps, twschum user files

### PR DESCRIPTION
**Pull request by twschum**
## Description
Adds two keyboard layouts for the DZ60 from kbdfans using a split spacebar layout. 
Adds Levinson (from keebio) layout that uses one half as a vertical numpad.
Adds custom functions in users/twschum

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* N/A

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/dz60/keymaps/twschum_b_4_10_simple/keymap.c [hunk 1]
```diff
@@ -0,0 +1,81 @@
+#include "dz60.h"
+
+/* DZ60 layout using following options (from layouts diagram on KBDfans):
+ *  - plate B (2.25u lshift)
+ *  - opt 4 (1.75, 1, 1 on rshift)
+ *  - opt 10 (2.75, 1.25, 2.25 on space, 5x1u bottom right keys)
+ * http://www.keyboard-layout-editor.com/#/gists/225f0f4dcf6671405f744fabe314627c
+ */
+
+#define MODS_CTRL_MASK  (MOD_BIT(KC_LSHIFT)|MOD_BIT(KC_RSHIFT))
```
Comment (drashna) [inline keyboards/dz60/keymaps/twschum_b_4_10_simple/keymap.c]: This isn't needed, actually. 

```suggestion
```

`MOD_MASK_CTRL` is actually globally defined and does the same thing. 

Also, this doesn't appear to be used at all, anyways.

File: keyboards/dz60/keymaps/twschum_b_4_10/keymap.c [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+ *  - plate B (2.25u lshift)
+ *  - opt 4 (1.75, 1, 1 on rshift)
+ *  - opt 10 (2.75, 1.25, 2.25 on space, 5x1u bottom right keys)
+ * http://www.keyboard-layout-editor.com/#/gists/225f0f4dcf6671405f744fabe314627c
+ */
+
+/* TODO: provide a standalone B_4_10 layout
+ */
+
+#undef KEYMAP
```
Comment (drashna) [inline keyboards/dz60/keymaps/twschum_b_4_10/keymap.c]: This line is not actually needed, and should be removed. 

```suggestion
```

File: keyboards/dz60/keymaps/twschum_b_4_10/keymap.c [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+#define KEYMAP( \
```
Comment (drashna) [inline keyboards/dz60/keymaps/twschum_b_4_10/keymap.c]: If you could, replace all of the `KEYMAP`s with `LAYOUT`.  We've switched terminologies, since `LAYOUT` is more accurate based on what is happening.

File: quantum/rgblight_list.h [hunk 4]
```diff
@@ -16,44 +16,46 @@
...
+#define HSV_GREEN         85, 255, 255
+#define HSV_SPRINGGREEN  106, 255, 255
+#define HSV_TURQUOISE    123,  90, 112
+#define HSV_TEAL         128, 255, 128
+#define HSV_CYAN         128, 255, 255
+#define HSV_AZURE        132, 102, 255
+#define HSV_BLUE         170, 255, 255
+#define HSV_PURPLE       191, 255, 255
+#define HSV_MAGENTA      213, 255, 255
+#define HSV_PINK         234, 128, 255
```
Comment (drashna) [inline quantum/rgblight_list.h]: Could this change be reverted? 

Specifically, we want to keep changes to core files separate from user or even keyboard related changes.  This is especially important if we ever need to roll back a commit. 

If you want, you can submit this change as a separate PR. 

Also, this change would be reverted as soon as it was committed, since we run a re-format pass after merging.

File: users/twschum/twschum.c [hunk 5]
```diff
@@ -0,0 +1,270 @@
+#include "twschum.h"
+
+// util for detecting simultanious shifts
+#define MODS_CTRL_MASK  (MOD_BIT(KC_LSHIFT)|MOD_BIT(KC_RSHIFT))
```
Comment (drashna) [inline users/twschum/twschum.c]: As above.  This doesn't need to be here, since we have `MOD_MASK_CTRL` defined globally.

File: users/twschum/twschum.c [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+/* Use RGB underglow to indicate layer
+ * https://docs.qmk.fm/reference/customizing-functionality
+ */
+// add to quantum/rgblight_list.h
+#ifdef RGBLIGHT_ENABLE
+static bool rgb_layers_enabled = true;
+static bool rgb_L0_enabled = false;
+
+uint32_t layer_state_set_user(uint32_t state) {
```
Comment (drashna) [inline users/twschum/twschum.c]:
```suggestion
layer_state_t layer_state_set_user(layer_state_t state) {
```

File: users/twschum/twschum.c [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    if (!rgb_layers_enabled) {
+        return state;
+    }
+    switch (biton32(state)) {
```
Comment (drashna) [inline users/twschum/twschum.c]:
```suggestion
    switch (get_highest_layer(state)) {
```

File: users/twschum/twschum.c [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+            { //
+                #if defined(__arm__)  // only run for ARM boards
+                    SEND_STRING(":dfu-util");
+                #elif defined(BOOTLOADER_DFU) // only run for DFU boards
+                    SEND_STRING(":dfu");
+                #elif defined(BOOTLOADER_HALFKAY) // only run for teensy boards
+                    SEND_STRING(":teensy");
+                #elif defined(BOOTLOADER_CATERINA) // only run for Pro Micros
+                    SEND_STRING(":avrdude");
+                #endif // bootloader options
```
Comment (drashna) [inline users/twschum/twschum.c]: All of this could be replaced with just: 
```
SEND_STRING(":flash");
```

File: keyboards/dz60/keymaps/twschum_b_4_10/keymap.c (hunk 3 above)
Comment (twschum) [reply keyboards/dz60/keymaps/twschum_b_4_10/keymap.c]: Will do!

File: users/twschum/twschum.c (hunk 8 above)
Comment (twschum) [reply users/twschum/twschum.c]: Oh awesome

**Comment by twschum** [comment]
I'm not sure what's going on with the lib/lufa changes. When I updated to latest qmk upstream, I had some trouble with the submodule stuff and think I took a version that was ahead of what's pinned in qml upstream? Not sure what to do to sync that.

File: quantum/rgblight_list.h (hunk 4 above)
Comment (twschum) [reply quantum/rgblight_list.h]: Will do

File: keyboards/dz60/keymaps/split_space_arrows/keymap.c [hunk 9]
```diff
@@ -0,0 +1,78 @@
+#include "dz60.h"
```
Comment (noroadsleft) [inline keyboards/dz60/keymaps/split_space_arrows/keymap.c]: This should be:

```suggestion
#include QMK_KEYBOARD_H
```

File: keyboards/dz60/keymaps/split_space_arrows/keymap.c [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+// Mac vs Linux/windows behaviour for media next/prev keys
+#define IS_MAC 1
+#ifdef IS_MAC
+#define MEDIA_FWD KC_MFFD
+#define MEDIA_RWD KC_MRWD
+#elif
+#define MEDIA_FWD KC_MNXT
+#define MEDIA_RWD KC_MPRV
+#define
+#endif
```
Comment (noroadsleft) [inline keyboards/dz60/keymaps/split_space_arrows/keymap.c]: `KC_MNXT` and `KC_MPRV` both work on macOS; if you assign those keycodes you don't need this logic at all.

File: keyboards/dz60/keymaps/split_space_arrows/rules.mk [hunk 11]
```diff
@@ -0,0 +1,7 @@
+# Build Options
+# override the options specified in dz60/rules.mk
+
+BOOTMAGIC_ENABLE = no	# Virtual DIP switch configuration(+1000)
```
Comment (noroadsleft) [inline keyboards/dz60/keymaps/split_space_arrows/rules.mk]: We're trying to get rid of these size impact numbers; they're seldom if ever accurate.

```suggestion
BOOTMAGIC_ENABLE = no	# Virtual DIP switch configuration
```

File: keyboards/dz60/keymaps/twschum_b_4_10/keymap.c [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+// Mac vs Linux/windows behaviour for media next/prev keys
+#define IS_MAC 1
+#ifdef IS_MAC
+#define MM_NEXT KC_MFFD
+#define MM_PREV KC_MRWD
+#elif
+#define MM_NEXT KC_MNXT
+#define MM_PREV KC_MPRV
+#define
+#endif
```
Comment (noroadsleft) [inline keyboards/dz60/keymaps/twschum_b_4_10/keymap.c]: Ditto here as above.

File: keyboards/dz60/keymaps/twschum_b_4_10/rules.mk [hunk 13]
```diff
@@ -0,0 +1,16 @@
+# Build Options
+# override the options specified in dz60/rules.mk
+BOOTMAGIC_ENABLE = no   # Virtual DIP switch configuration(+1000)
```
Comment (noroadsleft) [inline keyboards/dz60/keymaps/twschum_b_4_10/rules.mk]:
```suggestion
BOOTMAGIC_ENABLE = no   # Virtual DIP switch configuration
```

File: keyboards/dz60/keymaps/twschum_b_4_10/rules.mk [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+MOUSEKEY_ENABLE  = yes  # Mouse keys(+4700)
```
Comment (noroadsleft) [inline keyboards/dz60/keymaps/twschum_b_4_10/rules.mk]:
```suggestion
MOUSEKEY_ENABLE  = yes  # Mouse keys
```

File: keyboards/dz60/keymaps/twschum_b_4_10/rules.mk [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+EXTRAKEY_ENABLE  = yes  # Audio control and System control(+450)
```
Comment (noroadsleft) [inline keyboards/dz60/keymaps/twschum_b_4_10/rules.mk]:
```suggestion
EXTRAKEY_ENABLE  = yes  # Audio control and System control
```

File: keyboards/dz60/keymaps/twschum_b_4_10/rules.mk [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+CONSOLE_ENABLE   = no   # Console for debug(+400)
```
Comment (noroadsleft) [inline keyboards/dz60/keymaps/twschum_b_4_10/rules.mk]:
```suggestion
CONSOLE_ENABLE   = no   # Console for debug
```

File: keyboards/dz60/rules.mk [hunk 17]
```diff
@@ -21,8 +21,8 @@ CONSOLE_ENABLE = no	# Console for debug(+400)
 COMMAND_ENABLE = no    # Commands for debug and configuration
 SLEEP_LED_ENABLE = no  # Breathing sleep LED during USB suspend
 NKRO_ENABLE = yes		# USB Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
-BACKLIGHT_ENABLE = yes  # Enable keyboard backlight functionality
-AUDIO_ENABLE = no
-RGBLIGHT_ENABLE = yes
+BACKLIGHT_ENABLE = yes  # Enable per-key backlight LEDs
```
Comment (noroadsleft) [inline keyboards/dz60/rules.mk]: Please use the comments from the [AVR template](https://github.com/qmk/qmk_firmware/blob/1c9e4502181dc36ed9853dec4341765c87530240/quantum/template/avr/rules.mk#L27-L42).

```suggestion
BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality 
```

File: keyboards/dz60/rules.mk [hunk 18]
```diff
@@ -21,8 +21,8 @@ CONSOLE_ENABLE = no	# Console for debug(+400)
...
+AUDIO_ENABLE = no    # There is no speaker on this PCB
```
Comment (noroadsleft) [inline keyboards/dz60/rules.mk]:
```suggestion
AUDIO_ENABLE = no           # Audio output on port C6
```

File: keyboards/dz60/rules.mk [hunk 19]
```diff
@@ -21,8 +21,8 @@ CONSOLE_ENABLE = no	# Console for debug(+400)
...
+RGBLIGHT_ENABLE = yes # Enable the RGB underglow LEDs
```
Comment (noroadsleft) [inline keyboards/dz60/rules.mk]:
```suggestion
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
```

File: keyboards/keebio/levinson/keymaps/numpad/keymap.c [hunk 20]
```diff
@@ -0,0 +1,124 @@
+#include QMK_KEYBOARD_H
+#include "quantum.h"
```
Comment (noroadsleft) [inline keyboards/keebio/levinson/keymaps/numpad/keymap.c]: `QMK_KEYBOARD_H` includes `quantum.h` already.

```suggestion
```

File: users/twschum/xtonhasvim.c [hunk 21]
```diff
@@ -0,0 +1,630 @@
...
+  PRESS(KC_LSHIFT);
+    TAP(keycode);
+  RELEASE(KC_LSHIFT);
+}
+
+static inline void ALT(uint16_t keycode) {
+  PRESS(KC_LALT);
+    TAP(keycode);
+  RELEASE(KC_LALT);
+}
```
Comment (drashna) [inline users/twschum/xtonhasvim.c]: Honestly, none of this is needed.

`tap_code` allows for basic keycodes to be used, and `tap_code16` allofr for modded keycodes, (such as `tap_code16(LSFT(KC_I))`, or `tap_code16(LALT(KC_TAB))`

File: users/twschum/xtonhasvim.h [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef USERSPACE
+#define USERSPACE
+
+#include QMK_KEYBOARD_H
+#include "action_layer.h"
```
Comment (drashna) [inline users/twschum/xtonhasvim.h]: This line is not actually needed, and should be removed. 

```suggestion
```

File: users/twschum/xtonhasvim.h [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+  VIM_PERIOD, // to support indent/outdent
+  VIM_COMMA,  // and toggle comments
+  VIM_SHIFT, // avoid side-effect of supporting real shift.
+  VIM_ESC, // bookend
+  VIM_SAFE_RANGE // start other keycodes here.
+};
+
+bool process_record_vimlayer(uint16_t keycode, keyrecord_t *record);
+
+// NOTE: YOU MUST DEFINE THIS
```
Comment (drashna) [inline users/twschum/xtonhasvim.h]: Actually, this isn't true.  If you add the following to the c file above, you don't need to define this:
```c
__attribute__((weak))
uint8_t vim_cmd_layer(void) { return 0; }
```

Specifically, the "weak" attribute allows the function to be defined elsewhere, and gets overridden, so no conflicts occur.  In fact, QMK firmware makes heavy use of this (for instance, that's how `process_record_user` is handled. In fact, all of the `*_user` functions are handled this way.

File: users/twschum/xtonhasvim.c [hunk 24]
```diff
@@ -0,0 +1,593 @@
...
+                // tap_code16(LGUI(KC_RIGHT));
+                tap_code16(LCTL(KC_E));
+              } else {
+                tap_code(KC_RIGHT);
+              }
+              EDIT;
+              break;
+            case VIM_B:
+              register_code(KC_LALT);
+              register_code(KC_LEFT);
```
Comment (drashna) [inline users/twschum/xtonhasvim.c]: You can do;
```
register_code16(A(KC_LEFT));
```

**Comment by twschum** [comment]
Thanks, I'll condense those first!
